fix: network scene manager does not handle full scene paths

NoelStephensUnity opened the pull request:
[MTT-1286](https://jira.unity3d.com/browse/MTT-1286)
This allows users to use a full scene path when loading a scene via the NetworkSceneManager.
Modified NetworkSceneManagerTests to include testing this update.

NoelStephensUnity committed:
fix

This allows a user to use the full asset path to load a scene.

NoelStephensUnity committed:
test

this tests the fix to be able to use the full scene path to load a scene.

NoelStephensUnity committed:
refactor

Needed to make adjustments where it was the full scene path and name without the Assets and .unity.

NoelStephensUnity committed:
Merge branch 'develop' into fix/full-scene-path-NetworkSceneManager

NoelStephensUnity committed:
Merge branch 'develop' into fix/full-scene-path-NetworkSceneManager

NoelStephensUnity committed:
Merge branch 'develop' into fix/full-scene-path-NetworkSceneManager

NoelStephensUnity committed:
fix

fixing merge issue.

LukeStampfli reviewed (approved):
Looks good. So this will mean we will support everything but the solution where you use "Assets" and ".unity" right?

NoelStephensUnity committed:
Merge branch 'develop' into fix/full-scene-path-NetworkSceneManager

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -258,22 +263,37 @@ internal void EndSceneEvent(uint sceneEventId)
+        /// <returns>full path and scene name</returns>
+        internal string GetScenePathAndName(string scenePath)
+        {
+            return scenePath.Replace("Assets/", "").Replace(".unity", "");
```
Is this always true or could it be it happens to be true for the projects we set up?  Then I wonder if Unity has an API for this.  Like I see [this](https://docs.unity3d.com/ScriptReference/SceneManagement.Scene-path.html).  Is this where we don't have an actual scene to query, just paths?  I just want to challenge to make sure there isn't a more standard way of doing this

NoelStephensUnity replied:
I believe the Assets/ folder is always true for the scope of V1.0.0, when we get to addresables this all will have to change...but at that point we hopefully will be working with a SceneManagerAPI derived NetworkSceneManager so much of this code will no longer exist.

The scenePath is coming from the SceneUtility.GetScenePathByBuildIndex which will return the full path to the scene asset that will include the Assets/ and the <scene name>.unity extension.

mattwalsh-unity replied:
what would happen if a user had a scene named like this: `Assets/assets.unity/ArtAssets/someScene.unity`

`GetScenePathAndName` would rename that to `Assets/assets/Art/SomeScene.unity`

NoelStephensUnity replied:
This is handled and there is an additional test to verify that.  👍

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -509,6 +529,11 @@ private void SendSceneEventData(uint sceneEventId, ulong[] targetClientIds)
         /// <returns>true (Valid) or false (Invalid)</returns>
         internal bool IsSceneNameValid(string sceneName)
         {
+            if (sceneName.Contains("/"))
```
Yeah, again I start getting the jitters thinking this must be something we can more directly in the engine.  Then again I have not dug into it.   

Lastly, I assume this works on all platforms (e.g. \ vs. / )

NoelStephensUnity replied:
Yeah, the one to watch out for is the "\" as that typically requires "\\".  "/" is the better/more universal way to handle paths.

NoelStephensUnity committed:
Merge branch 'develop' into fix/full-scene-path-NetworkSceneManager

NoelStephensUnity committed:
refactor and fix

Fixing an issue with IsSceneNameValid
Including minor merge fixes.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -509,6 +529,22 @@ private void SendSceneEventData(uint sceneEventId, ulong[] targetClientIds)
         {
+            if (sceneName.Contains("/"))
+            {
+                if (sceneName.Contains(".unity"))
```
Again, I don't think it's illegal for a user to have a path that has `.unity` somewhere in it.  Maybe, this could be salvaged by only replacing it if the file ends with `.unity`, not just contains it

NoelStephensUnity replied:
This is addressed and a test was added to verify the various irregular paths and extension formats will be handled properly.

mattwalsh-unity reviewed (changes requested):
I would like to either just give up on trying to cleverly interpret these paths or find a more official way.  @LukeStampfli can you comment on how big a pain point this is for users?

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -258,22 +263,37 @@ internal void EndSceneEvent(uint sceneEventId)
+        /// </summary>
+        /// <param name="scenePath"></param>
+        /// <returns>full path and scene name</returns>
+        internal string GetScenePathAndName(string scenePath)
```
Please conspicuously mark this and similar string parsing trickery with "to be removed with scene v2.0, etc"

NoelStephensUnity replied:
Done!

NoelStephensUnity committed:
refactor

Added additional testing for various types of irregular but valid scene path names.
Renamed a SceneEventDataPoolSceneLoadingTest to SceneLoadingSceneEventDataStressTest in order to better reflect the test.

NoelStephensUnity committed:
Merge branch 'develop' into fix/full-scene-path-NetworkSceneManager

NoelStephensUnity committed:
style

Added additional request comments on string name and/or path manipulation and checking.

NoelStephensUnity committed:
Merge branch 'develop' into fix/full-scene-path-NetworkSceneManager

andrews-unity commented:
For my own sanity... why do we need to do the parsing? Like what are we trying to support here? 

So I see we are using the build index we have an API that takes name or a path and returns a build index: 

```
var valid = SceneUtility.GetBuildIndexByScenePath("SecondSceneToLoad");
var validAgain = SceneUtility.GetBuildIndexByScenePath($"Assets/Tests/Manual/SceneTransitioning/SecondSceneToLoad.unity");
var invalid = SceneUtility.GetBuildIndexByScenePath($"Assets/Tests/Manual/SceneTransitioning/SecondSceneToLoad");
```

Like we can pass in values to this API and it will return -1 if its an invalid path or and invalid scene I don't think we need to have all this validation code. Let the engine do that. 

Second why do we need to parse all this stuff?

NoelStephensUnity committed:
refactor

Replaced scene path checks with SceneUtility.GetBuildIndexByScenePath
Removed the additional test.
Removed the additional ScenePathsInBuild list.

Keeping the ScenesInBuild due to its wide usage, but this too could most likely be removed with additional modifications.

NoelStephensUnity committed:
Merge branch 'fix/full-scene-path-NetworkSceneManager' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into fix/full-scene-path-NetworkSceneManager

NoelStephensUnity committed:
fix test

Fixing the NetworkObjectParentingTests by adding the scene that is loaded to the scenes in build list in order to account for the recent changes.

NoelStephensUnity commented:
This became obsolete with [PR-1256](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1256)
